extension/observer/ecsobserver/factory.go

```diff
@@ -42,7 +45,13 @@ func createDefaultConfig() config.Extension {
 
 func createExtension(ctx context.Context, params component.ExtensionCreateSettings, cfg config.Extension) (component.Extension, error) {
 	sdCfg := cfg.(*Config)
-	sd, err := NewDiscovery(*sdCfg, ServiceDiscoveryOptions{Logger: params.Logger})
+	opt := serviceDiscoveryOptions{Logger: params.Logger}
+	// Only for test
+	fetcher := ctx.Value(ctxFetcherOverrideKey)
+	if fetcher != nil {
+		opt.FetcherOverride = fetcher.(*taskFetcher)
+	}
+	sd, err := newDiscovery(*sdCfg, opt)
```

Comment:
If it is for test purpose only, can we move this piece of code in unit test only instead of having it here ?

Reply:
That's for testing create extension, I could pass the ctx to `newDiscovery` and move the logic there. But still it's not test only and always get executed.

Reply:
I agree with Min that this doesn't feel right.  I'm wary of the amount of "test only" code.  If the fetcher is a dependency then allow dependency injection through configuration or initializer parameters or something, and have tests provide a mock dependency during testing.  Having special cases doesn't seem like the right way to do this.

Reply:
Moved to `Config` as there are just three parameters for creating extension `func createExtension(ctx context.Context, params component.ExtensionCreateSettings, cfg config.Extension)`, originally I don't want to update `Config` as it's more a POJO like struct, but it's a bit type safer than using context and that Config struct is not expected to be used outside package. Though I don't remember reflection based Unmarshal works with unexported struct.

A second thought, I could just have an extra `createExtensionWithFetcher` for test and avoid updating `Config`.

Reply:
Done in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/3785/commits/87f107b15e4b3a2d1efd9e3b68c648d58191ac2a